File: python/BOJ/BOJ11404.py
# https://www.acmicpc.net/problem/11404
# 플로이드

import sys

IMPOSSIBLE = 987654321

n = int(sys.stdin.readline())
m = int(sys.stdin.readline())

dist = [[IMPOSSIBLE for loop in range(n) ] for loop in range(n)]
# graph = [[IMPOSSIBLE for loop in n] for loop in n]


for loop in range(n):
    dist[loop][loop] = 0

for loop in range(m):
    st, en, cost = map(int, sys.stdin.readline().split())
    st -= 1
    en -= 1
    dist[st][en] = min(cost, dist[st][en])

for k in range(n):
    for i in range(n):
        for j in range(n):
            if (dist[i][k] == IMPOSSIBLE) or (dist[k][j] == IMPOSSIBLE):
                continue
            if dist[i][j] > (dist[i][k] + dist[k][j]):
                dist[i][j] = dist[i][k] + dist[k][j]

# print
for i in range(n):
    for j in range(n):
        if dist[i][j] == IMPOSSIBLE:
            sys.stdout.write(str(0) + ' ')
        else:
            sys.stdout.write(str(dist[i][j]) + ' ')

    sys.stdout.write(str('\n'))

File: BOJ/BOJ11404.java
import java.io.*;
import java.util.*;


public class Main {

    public static final int IMPOSSIBLE = 100_000_000;
    public static int N, M;
    public static int[][] dist;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        M = Integer.parseInt(br.readLine());

        dist = new int[N][N];

        for (int i = 0; i < N; ++i) {
            Arrays.fill(dist[i], IMPOSSIBLE);
        }

        for (int loop = 0; loop < M; ++loop) {
            String[] edgeInfo = br.readLine().split(" ");
            int st = Integer.parseInt(edgeInfo[0]);
            int en = Integer.parseInt(edgeInfo[1]);
            int cost = Integer.parseInt(edgeInfo[2]);

            st--; en--;
            dist[st][en] = Math.min(dist[st][en], cost);
        }

        for (int k = 0; k < N; ++k) {
            for (int i = 0; i < N; ++i) {
                for (int j = 0; j < N; ++j) {
                    if (i == j) continue;

                    if (dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                int val = dist[i][j];

                if (val == IMPOSSIBLE) {
                    sb.append(0 + " ");
                } else {
                    sb.append(val + " ");
                }
            }
            sb.append('\n');
        }

        System.out.println(sb);
    }

}
